src/endf.cpp

```diff
@@ -0,0 +1,261 @@
+  std::string func_type;
+  read_attribute(dset, "type", func_type);
+  std::unique_ptr<Function1D> func;
+  if (func_type == "Tabulated1D") {
+    func = std::make_unique<Tabulated1D>(dset);
+  } else if (func_type == "Polynomial") {
+    func = std::make_unique<Polynomial>(dset);
+  } else if (func_type == "CoherentElastic") {
+    func = std::make_unique<CoherentElasticXS>(dset);
+  } else if (func_type == "IncoherentElastic") {
+    func = std::make_unique<IncoherentElasticXS>(dset);
+  } else {
+    throw std::runtime_error{"Unknown function type " + func_type +
+      " for dataset " + object_name(dset)};
+  }
+  close_dataset(dset);
+  return func;
+}
+
+//==============================================================================
+// Polynomial implementation
+//==============================================================================
+
+Polynomial::Polynomial(hid_t dset)
+{
+  // Read coefficients into a vector
+  read_dataset(dset, coef_);
+}
+
+double Polynomial::operator()(double x) const
+{
+  // Use Horner's rule to evaluate polynomial. Note that coefficients are
+  // ordered in increasing powers of x.
+  double y = 0.0;
+  for (auto c = coef_.crbegin(); c != coef_.crend(); ++c) {
+    y = y*x + *c;
+  }
+  return y;
+}
+
+//==============================================================================
+// Tabulated1D implementation
+//==============================================================================
+
+Tabulated1D::Tabulated1D(hid_t dset)
+{
+  read_attribute(dset, "breakpoints", nbt_);
+  n_regions_ = nbt_.size();
+
+  // Change 1-indexing to 0-indexing
+  for (auto& b : nbt_) --b;
+
+  std::vector<int> int_temp;
+  read_attribute(dset, "interpolation", int_temp);
+
+  // Convert vector of ints into Interpolation
+  for (const auto i : int_temp)
+    int_.push_back(int2interp(i));
+
+  xt::xarray<double> arr;
+  read_dataset(dset, arr);
+
+  auto xs = xt::view(arr, 0);
+  auto ys = xt::view(arr, 1);
+
+  std::copy(xs.begin(), xs.end(), std::back_inserter(x_));
+  std::copy(ys.begin(), ys.end(), std::back_inserter(y_));
+  n_pairs_ = x_.size();
+}
+
+double Tabulated1D::operator()(double x) const
+{
+  // find which bin the abscissa is in -- if the abscissa is outside the
+  // tabulated range, the first or last point is chosen, i.e. no interpolation
+  // is done outside the energy range
+  int i;
+  if (x < x_[0]) {
+    return y_[0];
+  } else if (x > x_[n_pairs_ - 1]) {
+    return y_[n_pairs_ - 1];
+  } else {
+    i = lower_bound_index(x_.begin(), x_.end(), x);
+  }
+
+  // determine interpolation scheme
+  Interpolation interp;
+  if (n_regions_ == 0) {
+    interp = Interpolation::lin_lin;
+  } else {
+    interp = int_[0];
+    for (int j = 0; j < n_regions_; ++j) {
+      if (i < nbt_[j]) {
+        interp = int_[j];
+        break;
+      }
+    }
+  }
+
+  // handle special case of histogram interpolation
+  if (interp == Interpolation::histogram) return y_[i];
```

Comment:
Optional: Given this is short-circuiting the function, it may make more sense to move this up.  I would recommend including this in the `if` block starting on line `170`, specifically as the first condition.